hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieParquetRealtimeInputFormat.java

```diff
@@ -197,10 +197,27 @@ private static synchronized Configuration addRequiredProjectionFields(Configurat
     return configuration;
   }
 
+  /**
+   * Hive will append read columns' ids to old columns' ids during getRecordReader. In some cases, e.g. SELECT COUNT(*),
+   * the read columns' id is an empty string and Hive will combine it with Hoodie required projection ids and becomes
+   * e.g. ",2,0,3" and will cause an error. This method is used to avoid this situation.
+   */
+  private static synchronized Configuration cleanProjectionColumnIds(Configuration conf) {
```

Comment:
Is there a reason for using synchronized ? (Is this for non Hive on MR based jobs ?)

Reply:
Good question. Actually I am not sure about this. But I find that HoodieParquetRealtimeInputFormat::addRequiredProjectionFields method is synchronized. I guess this method should be similar with that.

Reply:
Like for Spark, multiple tasks run in the same executor. I think this could be a use case.

Reply:
As you said, the weird comma is added in the ```HiveInputFormat.java``` and then it directly calls getRecordReader from ```HoodieParquetRealtimeInputFormat.java```. I didn't see a way to do this check even earlier unless we do it in the Hive code.

Reply:
@zhedoubushishi : You can synchronize on the passed conf object instead of static synchronization which becomes a global lock at the JVM level. 

You can do something like 
synchronized(conf) {
 ....
}
inside your cleanProjectionColumnIds.

Reply:
Otherwise, looks ok.

Reply:
> @zhedoubushishi : You can synchronize on the passed conf object instead of static synchronization which becomes a global lock at the JVM level.
> 
> You can do something like
> synchronized(conf) {
> ....
> }
> inside your cleanProjectionColumnIds.

That make sense. Code changes are done.